### ISSUE TITLE
DynamicObject models not rendered correctly

### ISSUE THREAD

**Issue by freeranger** [domaindrivendev/Swashbuckle#412]
Hi,

Objects inheriting from DynamicObject end up rendered as empty models.

For example If I have a domain model:
```
    public class FooBar : DynamicObject
    {
        public string Foo { get; set; }
        public int Bar { get; set; }
    }
```
It is rendered as "`{}`"

Obviously I can't expect the dynamic elements of such an object to be rendered, but the non dynamic elements should be.

**Comment by domaindrivendev** [comment]
I'm actually curious to know if WebApi itself (regardless of Swashbuckle) serializes your FooBar object in the way you think. Trying out your example above, it looks to me that the behavior of the default JsonSerializer is to NOT include any of the properties from the base class

**Comment by domaindrivendev** [comment]
Closing due to inactivity. Feel free to re-open if you have more information around the question above

**Comment by freeranger** [comment]
Hi,
Sorry, was away - hence the inactivity :)

If I mark my FooBar object as  [JsonObject] then it is serialized as expected through the webapi - ie, as { "foo": "Fooey", "bar": 123 } for example.
Swashbuckle however still renders this as just "{}"

**Comment by freeranger** [comment]
hmm...I don't seem to be able to re-open this issue - just adding a comment isn't enough and there are no suitable links/buttons....

**Comment by domaindrivendev** [comment]
Ok gotcha - that's what I was missing. I think this should be a relatively straightforward fix. Will mark it down for the next release milestone

**Comment by ttarney** [comment]
To get around I added a class that inherits from DYnamicObject and then one level of indirection lower.  Seems that if there is a direct relationship to DynmaicObject it doesn't work.

That took care of it.

Thanks

**Comment by domaindrivendev** [comment]
I'm just revisiting this issue and can't repro on the latest version (5.2.2). Can you confirm this is still an issue and if so let me know which version of Newtonsoft you're using?

**Comment by domaindrivendev** [comment]
Have to close again due to inactivity. You know the drill - I can't repro but please reopen if you're still seeing this issue on latest version

**Comment by ttarney** [comment]
Yes it is goodThanksSent from my Verizon Wireless 4G LTE DROIDOn Jan 12, 2016 8:06 PM, domaindrivendev <notifications@github.com> wrote:Have to close again due to inactivity. You know the drill - I can't repro but please reopen if you're still seeing this issue on latest version

**Comment by freeranger** [comment]
Apologies for not getting back to you - here: https://github.com/freeranger/SwashbuckleDynamicObject is a minimal project that demonstrates the problem.
It is using Swashbuckle 5.3.2 and Newtonsoft.Json version 8.0.3 (also tested with 6.00..4 and the behaviour is the same)

if you run up the solution and navigate to `/swagger` you can see the docs for `foobar/get` - the model generated is `{}`.

Edit the `FooBar` class and add the `[JsonObject]` attribute to the class and it generates the expected model:
```
{
  "Foo": "string",
  "Bar": 0
}
```
Remove `[JsonObject]` and stop inheriting from `DynamicObject` and you also get the correct model.
Hopefully you will be able to reproduce now?

**Comment by freeranger** [comment]
@domaindrivendev - Have you managed to recreate the problem with the sample project I supplied?

**Comment by cobysy** [comment]
I managed to get it showing a schema and example with this workaround:

```
public class MissingSchemaOnDynamicOperationFilter : IOperationFilter
{
    public void Apply(Operation operation, SchemaRegistry schemaRegistry, ApiDescription apiDescription)
    {
        ResponseTypeAttribute attr = apiDescription.ActionDescriptor.GetCustomAttributes<ResponseTypeAttribute>().FirstOrDefault();
        if (attr == null)
            return;

        Schema referenceToSchema = schemaRegistry.GetOrRegister(attr.ResponseType);

        Response responseValue = operation.responses["200"];
        if (responseValue.schema == null)
            responseValue.schema = referenceToSchema;

    }
}
```

Of course you need to decorate your method with the ResponseTypeAttribute.
Also I have my own IDynamicMetaObjectProvider implementation in order to get the json serializer working (largely based on https://loosexaml.wordpress.com/2011/01/01/wcf-serialization-of-dlr-dynamic-types/)

**Comment by cobysy** [comment]
Also you need this:

```
// Customize JsonContractResolver so that Swashbuckle understands dynamic objects
httpConfiguration.Formatters.JsonFormatter.SerializerSettings.ContractResolver = new CustomJsonContractResolver(jsonContractResolver);
```

```
public class CustomJsonContractResolver : IContractResolver
{
    private readonly IContractResolver _underlyingResolver;

    public CustomJsonContractResolver(JsonContractResolver underlyingResolver)
    {
        _underlyingResolver = underlyingResolver;
    }

    public JsonContract ResolveContract(Type type)
    {
        JsonContract resolveContract = _underlyingResolver.ResolveContract(type);

        if (type == typeof (RegisterUpdateStartedRequest))
        {
            // Convert JsonDynamicObject to JsonObjectContract that SwashBuckle's SchemaRegistry.CreateDefinitionSchema() knows how to handle
            // https://github.com/domaindrivendev/Swashbuckle/blob/f5f3a187e1648e818778eddbdb4e1b6e35b6d83d/Swashbuckle.Core/Swagger/SchemaRegistry.cs
            JsonObjectContract jsonObjectContract = new JsonObjectContract(type);
            foreach (JsonProperty jsonProperty in ((JsonDynamicContract) resolveContract).Properties)
            {
                jsonObjectContract.Properties.Add(jsonProperty);
            }
            return jsonObjectContract;
        }

        return resolveContract;
    }
}
```

**Comment by freeranger** [comment]
Thanks.  I can get the desired output by decorating with `[JsonObject]` which is rather less work, but I think it should work without even that and this is a bug.
I'm not sure if @domaindrivendev still sees these comments on closed issues for him to comment on it :/

**Comment by freeranger** [comment]
@domaindrivendev - this issue is marked as "waiting for response" but I have already given response, with a minimal project demonstrating the problem - is there anything else you need?

**Comment by domaindrivendev** [comment]
As u can see I've already reopened the ticket. When I'll actually get to it is another story. To be honest, it's lower on the list of priorities, especially if there's a workaround.

Right now, my free time (which is minimal) Is dedicated to a version that works with next-gen .Net - AspNet Core.

**Comment by freeranger** [comment]
@domaindrivendev - Sure no prob - just wanted to make sure you weren't waiting on me for any more input.
There is a "workaround" as in it is possible to get the desired output if you "pollute" the object model with extra attributes - provided of course you have access to the source code for the models, but for me it's not really ideal.
If you could give some pointers to where the problem may lie, maybe I will find some time to take a look and have a go myself - though no promises as I am fully aware of the difficulties of finding free time to work on open source projects :}

Thanks for re-opening anyway, at least it's not forgotten!

**Comment by HeroXXiv** [comment]
Hi,

For me, I'm using [`Westwind.Utilities.Expando`](https://github.com/RickStrahl/Westwind.Utilities/blob/7d5d76f3b415f9e43ebd558a467fb9cf8e5de24a/Westwind.Utilities/SupportClasses/Expando.cs) as my base dynamic model which supports both compiletime and runtime properties, and generating openapi document from business logic services directly, so i am able to invoke `SchemaRegistry` with my code.

According to [`SchemaRegistry.CreateInlineSchema`](https://github.com/domaindrivendev/Swashbuckle/blob/5489aca0d2dd7946f5569341f621f581720d4634/Swashbuckle.Core/Swagger/SchemaRegistry.cs#L88), a class derived from `DynamicObject` will results `JsonDynamicContract` but `JsonObjectContract` if marked with `[JsonObject]`. So IMO I can 

1. add `else if (jsonContract is JsonDynamicContract)` check and return a schema just like `CreateObjectSchema` with `AdditionalProperties` as well, or
2. update [`CustomTypeMappings` check logic](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/28e8673e2579b406bc5120b89589aee3b5b79341/src/Swashbuckle.AspNetCore.SwaggerGen/Generator/SchemaRegistry.cs#L89) with inheritance support, then I can add a mapping of Expando class and which works with all derived model classes, or
3. add a `SchemaFilter` to modify the current object schema returned. (may not change any logic in `SchemaRegistry` but need some cross/cycle invokes between filter and registry, such as `CreatePropertySchema` and `referencedTypes`)

**Comment by MikeGriffinReborn** [comment]
Yes, this is a bug issue for us too. There is absolutely nothing wrong with having a class inherit from DynamicObject, it serialized just fine in JSON and XML, I added a new bug on this today

https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/1314
